[pytorch] get rid of cpp_type_str from pybind codegen

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46978 [pytorch] continue to rewrite gen_python_functions.py with typed models
* **#46977 [pytorch] get rid of cpp_type_str from pybind codegen**
* #46976 [pytorch] simplify tensor options logic in pybinding codegen
* #46243 [pytorch] add script to run all codegen

Clean up a few TODOs in the new python binding codegen.
Get rid of the _simple_type() hack and the uses of cpp_type_str.
Now python argument type strings and PythonArgParser unpacking methods
are directly generated from the original Type model.

Differential Revision: [D24589209](https://our.internmc.facebook.com/intern/diff/D24589209)

ljk53 committed:
[pytorch] get rid of cpp_type_str from pybind codegen

Clean up a few TODOs in the new python binding codegen.
Get rid of the _simple_type() hack and the uses of cpp_type_str.
Now python argument type strings and PythonArgParser unpacking methods
are directly generated from the original Type model.

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] get rid of cpp_type_str from pybind codegen"


Clean up a few TODOs in the new python binding codegen.
Get rid of the _simple_type() hack and the uses of cpp_type_str.
Now python argument type strings and PythonArgParser unpacking methods
are directly generated from the original Type model.

[ghstack-poisoned]

ezyang commented on tools/codegen/api/python.py:
```diff
@@ -817,27 +832,15 @@ def arg_parser_output_expr(
-        unpack_with_size = UNPACK_WITH_SIZE_METHODS[typename].format(a.size)
-        expr = f'_r.{unpack_with_size}({arg_index})'
+    if a.name != 'out' and la is not None and isinstance(la.argument, Argument):
+        arg_type = la.argument.type
```
This still seems a bit more complicated than it should be

ljk53 replied:
> This still seems a bit more complicated than it should be

Yes, I figured we don't need use la.argument.type anymore as now we can directly derive it from 'local.use_c10_dispatcher().dispatcher_uses_new_style()' :)

Deleted it and the long comment above!

ezyang reviewed (approved):
Thank you!

ljk53 committed:
Update on "[pytorch] get rid of cpp_type_str from pybind codegen"


Clean up a few TODOs in the new python binding codegen.
Get rid of the _simple_type() hack and the uses of cpp_type_str.
Now python argument type strings and PythonArgParser unpacking methods
are directly generated from the original Type model.

Differential Revision: [D24589209](https://our.internmc.facebook.com/intern/diff/D24589209)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] get rid of cpp_type_str from pybind codegen"


Clean up a few TODOs in the new python binding codegen.
Get rid of the _simple_type() hack and the uses of cpp_type_str.
Now python argument type strings and PythonArgParser unpacking methods
are directly generated from the original Type model.

Differential Revision: [D24589209](https://our.internmc.facebook.com/intern/diff/D24589209)

[ghstack-poisoned]
